src/Microsoft.ML.Parquet/ParquetLoader.cs

```diff
@@ -390,9 +390,21 @@ public Cursor(ParquetLoader parent, Func<int, bool> predicate, IRandom rand)
                     Columns = _loader._columnsLoaded.Select(i => i.Name).ToArray()
                 };
 
-                int numBlocks = (int)Math.Ceiling(((decimal)parent.GetRowCount() / _readerOptions.Count));
-                int[] blockOrder = _rand == null ? Utils.GetIdentityPermutation(numBlocks) : Utils.GetRandomPermutation(rand, numBlocks);
-                _blockEnumerator = blockOrder.GetEnumerator();
+                try
+                {
+                    int numBlocks = checked((int)Math.Ceiling(((decimal)parent.GetRowCount() / _readerOptions.Count)));
+                    int[] blockOrder = _rand == null ? Utils.GetIdentityPermutation(numBlocks) : Utils.GetRandomPermutation(rand, numBlocks);
+                    _blockEnumerator = blockOrder.GetEnumerator();
+                }
+                catch (Exception e)
+                {
+                    if (e is OutOfMemoryException || e is OverflowException)
```

Comment:
Maybe I'd say one is a refinement of the other. The size can be so large that either (1) it can't fit in an int or (2) it does fit in an int but that's moot since we can't allocate an `int[]` array that large in .NET anyway. So I'd like the error message to be the same.

That said I still agree with your suggestion of separating the two out... we know exactly which line would throw which exception, and I'd prefer the slightly move verbose two try statements *because* we have that expectation, since it clarifies to the reader where we expect the two distinct failure cases of the code above to happen, possibly.

Also FYI, just a quick style note for the future: even if we let this code stand as is, we could convert this to `catch (Exception e) when (e is OutOfMemoryException || e is OverflowException)`. This simplifies things enormously. (This very useful feature was added in, I think, C# 6 but I wouldn't stake my life on it.)

---
In reply to: [186318417](https://github.com/dotnet/machinelearning/pull/12#discussion_r186318417) [](ancestors = 186318417)